spotty/spotty/Data/Network/Auth/AuthManager.swift

```diff
@@ -0,0 +1,165 @@
+        guard let expirationDate = accessTokenExpirationDate else { return false }
+        let currentDate = Date()
+        let fiveMinutes: TimeInterval = 300
+        return currentDate.addingTimeInterval(fiveMinutes) >= expirationDate
+    }
+    
+    private var accessToken: String? {
+        guard let data = KeychainManager.read(
+            service: KeychainManager.Services.accessToken,
+            account: KeychainManager.Accounts.spotify)
+        else { return nil }
+        return String(data: data, encoding: .utf8)!
+    }
+    
+    private var refreshToken: String? {
+        guard let data = KeychainManager.read(
+            service: KeychainManager.Services.refreshToken,
+            account: KeychainManager.Accounts.spotify)
+        else { return nil }
+        return String(data: data, encoding: .utf8)!
+    }
+    
+    private var accessTokenExpirationDate: Date? {
+        guard let data = KeychainManager.read(
+            service: KeychainManager.Services.expirationDate,
+            account: KeychainManager.Accounts.spotify)
+        else { return nil }
+        let retrievedTimestamp = data.withUnsafeBytes { $0.load(as: Double.self) }
+        return Date(timeIntervalSinceReferenceDate: retrievedTimestamp)
+    }
+    
+    // MARK: - Exposed methods
+    
+    /// Exchange the Code from the API for an Access Token and cache the response
+    /// - Parameters:
+    ///   - code: The Code return by the external API as part of the OAuth2 Code flow.
+    ///   - completion: Handle successfull or unsuccessfull exchange.
+    ///
+    ///  The API Response that needs to be cached is represented by an ``AuthResponse`` model in this application.
+    ///  It contains multiple properties including Access Token, Refresh Token and Expiration Date
+    func exchangeCodeForAccessToken(code: String, completion: @escaping (Bool) -> Void) {
+        guard let request = apiDelegate.createTokenExchangeRequest(with: code) else { return }
+        
+        Network.performRequest(urlRequest: request) { [weak self] (result: Result<AuthResponse, Error>) in
+            switch result {
+            case .success(let response):
+                self?.cacheTokens(result: response)
+                completion(true)
+            case .failure(let error):
+                print(error.localizedDescription)
+                completion(false)
+            }
+        }
+    }
+    
+    /// Exchange the Refresh Token for an Access Token and cache the response.
+    /// - Parameter completion: Handle successfull or unsuccessfull exchange.
+    ///
+    /// The API Response that needs to be cached is represented by an ``AuthResponse`` model in this application.
+    /// It contains multiple properties including Access Token, Refresh Token and Expiration Date
+    func refreshTokenIfNeeded(completion: @escaping (Bool) -> Void) {
+        guard !isRefreshingToken, let refreshToken = refreshToken else { return }
```

Comment:
I suggest to move every statement to new line.

Reply:
done